### DISCUSSION TITLE
vi.mocked returns a mock that I can't use without a typescript error, even though my test runs exactly as I expect.

### DISCUSSION THREAD

**Discussion by doei** [vitest-dev/vitest#3193] [Q&A]
I have a function that I'm trying to mock, it calls an api and returns a user.
```typescript
export const getUser = async ({ userId }: { userId: string }): Promise<ApiUser> => {
	return await apiCall({ path: `user/${userId}`, method: 'GET' });
};
```

In my setup.ts, I am mocking the relevant module like so:
```typescript
vi.mock('../../src/services/apiClient', async () => {
	return { getUser: vi.fn() };
});
```

And in my test, I use `vi.mocked` to get the correct mocked function types: 

```typescript
import * as apiClient from '../../../src/services/apiClient';

const apiGetClientMock = vi.mocked(apiClient.getUser, { partial: true });

test("calls the api", () => {
	// the following will give me a typescript error
	apiGetClientMock.mockResolvedValueOnce({ id: "fake_user_id" });
})
```

As you can see I am using the `partial` option, because I am not interested in mocking the `ApiUser` entirely.

However, typescript now expects me to pass an argument of type `Partial<Promise<ApiUser>>'`, which is a bit confusing because `mockResolvedValue` should be trying to mock the _resolved_ value here (as the name suggests).

I also have this problem when using `mockImplementation` like this

```typescript
apiGetClientMock.mockImplementation(async () => ({ id: "fake_user_id" }));
```

The typescript error now reads: Type '{ id: string; }' has no properties in common with type 'Partial<Promise<ApiUser>>', but again that's not very useful.

Other potentially relevant information:
- When not using the `partial` option, it all behaves like I would expect.
- My test passes without issue and debugging the test shows me that the function and its implementation (or return value) are mocked as expected.

How can I mock an async function like this and have the mock return a resolved partial in a typesafe way ?

**Comment by sheremet-va** [comment]
There is no API in Vitest that will allow you to partially pass data to a mocked function. I would recommend creating a "dummy" object and pass it down in those cases. It is generally not a good idea to rely on implementation details. Treat the function as a black box where you know that it requires all these attributes. 

If you are not convinced, you can create your own helper and cast the type to `any`.

**Reply by doei** [reply]
I want to partially mock the return value of the function, not pass data partially.
I agree that not relying on implementation details seems like sound advice and if that's the choice made by vitest then all good, I'll take it.
But then why have that partial option at all ? Because this option _will_ allow you to partially mock the return value of a function if it's not a promise.

**Comment by sheremet-va** [comment]
There is no API in Vitest that will allow you to partially pass data to a mocked function. I would recommend creating a "dummy" object and pass it down in those cases. It is generally not a good idea to rely on implementation details. Treat the function as a black box where you know that it requires all these attributes. 

If you are not convinced, you can create your own helper and cast the type to `any`.
